### DISCUSSION TITLE
Body as dependency

### DISCUSSION THREAD

**Discussion by jochman** [fastapi/fastapi#6082] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Depends, Body
from pydantic import BaseModel

def validate_body(body: str = Body()):
  assert body == "body"

app = FastAPI(
  dependencies=[Depends(validate_body)]
)

class Model(BaseModel):
  var: str

@app.post("/")
async def just_runs(model: Model):
  return {"success": True}
```

### Description

I need to use the body in the authorization method, a method seems in services like [Hubspot](https://developers.hubspot.com/docs/api/webhooks/validating-requests).

The code above runs but the `body` parameter in `validate_body` func is string.

I've tried to get the body directly from the request with:

```python
async def validate_body(request: Request):
  body = await request.body()
  assert body == "body"
```

But the problem is that after the body is resolved it is not presented in the request object, therefore missing from the API.

### Operating System

Linux, macOS

### FastAPI Version

0.92.1

### Python Version

3.11

**Comment by ebottos94** [comment]
Hi @jochman, 

is not clare what you want to do. If your intent is validate body paramether, the validation is built in with pydantic. 
If your intent is validate the body if it is present (and use the same route if it isn't present) you can do simply something like this : 

```python
from fastapi import FastAPI, Body
from pydantic import BaseModel

app = FastAPI()

class Model(BaseModel):
  var: str 

@app.post("/")
async def just_runs(model: Model = Body(None)):
    return {"success": True}
```

**Reply by jochman** [reply]
I need to authorize the body with a function. So I have a dependency (just like a jwt token) and i need to check it against the body.

the authorization method is:

```python
import hashlib
import os 

secret_key = os.environ["SECRET_KEY"]

def validate(body: str, signature: str):
  message_to_validate = secret key + body
  create_signature = hashlib.sha256(meesage_to_validate.encode("utf-8")).hexdigest()
  assert create_signature == signature
```

So I need to use the body in the authorization method (and I need to use the body data).

I wanted to create the following file but cant:

```python
from fastapi import FastAPI, Body, Header, Depends
from pydantic import BaseModel

secret_key = os.environ["SECRET_KEY"]

async def validate_signature(
  signature: str = Header()
  body: str = Body()
):
  message_to_validate = secret key + body
  create_signature = hashlib.sha256(meesage_to_validate.encode("utf-8")).hexdigest()
  assert create_signature == signature


app = FastAPI(dependencies=[Depends(validate_signature)])

class Model(BaseModel):
  var: str 

@app.post("/")
async def just_runs(model: Model):
    return {"success": model.var}

**Reply by ebottos94** [reply]
@jochman if the value to validate is formed by header + value in body in my opinion a good way to procede is this : 

```python
from fastapi import FastAPI, Header
from pydantic import BaseModel

secret_key = "my_key"

app = FastAPI()

class Model(BaseModel):
  var: str 

def validate_signature(signature: str, var: str):
  #here write your validation
  message_to_validate = secret_key + var
  if message_to_validate == signature :
    return True
  else :
    return False

@app.post("/")
async def just_runs(model: Model, signature: str = Header()):
    print(signature)
    if validate_signature(signature, model.var) :
        return {"success": True}
    else :
       return {"success": False}
```

**Reply by jochman** [reply]
It is built from the whole body. So I can't use any processing on the body as it will change the hash value.

**Reply by ebottos94** [reply]
What do you mean with "whole body"? Post an example.

**Reply by jochman** [reply]
> What do you mean with "whole body"? Post an example.

I've posted an example

```python
from fastapi import FastAPI, Body, Header, Depends
from pydantic import BaseModel

secret_key = os.environ["SECRET_KEY"]

async def validate_signature(
  signature: str, # this is a header
  body: str # This is the body, whole document as sent to the api
):
  message_to_validate = secret key + body
  create_signature = hashlib.sha256(meesage_to_validate.encode("utf-8")).hexdigest()
  assert create_signature == signature


app = FastAPI(dependencies=[Depends(validate_signature)])

class Model(BaseModel):
  var: str 

@app.post("/")
async def just_runs(model: Model):
    return {"success": model.var}
```

I need to use it as Hubspot authentication https://developers.hubspot.com/docs/api/webhooks/validating-requests

**Reply by ebottos94** [reply]
try with this : 

```python
from fastapi import FastAPI, Header, Body
from pydantic import BaseModel
from typing import Any

secret_key = "my_key"

app = FastAPI()

class Model(BaseModel):
  var: str 

def validate_signature(signature: str, var: str):
  message_to_validate = secret_key + var
  if message_to_validate == signature :
    return True
  else :
    return False

@app.post("/")
async def just_runs(body: Any = Body(...), signature: str = Header()):
    if validate_signature(signature, body) :
        return {"success": True}
    else :
       return {"success": False}
```

**Comment by nymous** [comment]
Hello @jochman,

What issue did you have with the `await request.body()`?
I suggested this approach to a colleague when they wanted to validate a GitHub webhook payload, it seems to be the same idea here.

This code works fine for me, I can print the request body in the dependency (exactly as it is received) and still access it in an endpoint function:
```py
from fastapi import FastAPI, Header, Depends, Request, Security
from fastapi.security import APIKeyHeader
from pydantic import BaseModel

SignatureHeader = APIKeyHeader(name="X-Hubspot-Signature-256")

async def validate_signature(*, request: Request, hubspot_signature: str = Security(SignatureHeader)):
    body = await request.body()
    print(hubspot_signature)
    print(body)

app = FastAPI(dependencies=[Depends(validate_signature)])

class Model(BaseModel):
      var: str

@app.post("/")
async def just_runs(model: Model):
    return {"success": model.var}

```

Running FastAPI 0.92.0, starlette 0.25.0, uvicorn 0.20.0, Python 3.10.9 on Linux.

*(For reference, the code used to check GitHub payload signature: https://github.com/Tioborto/raddar/blob/main/raddar/core/security.py)*

**Reply by jochman** [reply]
after doing `await request.body)` in the dependency, the model was empty.

**Comment by YuriiMotov** [comment, marked as answer]
Problem with empty `request.body` was fixed in Starlette [0.28.0](https://github.com/encode/starlette/releases/tag/0.28.0)
